### PR TITLE
expression: make count distinct multi column aware of new collation

### LINKED ISSUE #27091: count distinct with multiple column return wrong result under new collation
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
```
mysql> create table t2(id int, v1 char(10) collate utf8mb4_general_ci, v2 char(10) collate utf8mb4_bin);
Query OK, 0 rows affected (0.01 sec)

mysql> insert into t2 values(1,'a','A'),(2,'A','A'),(3,'a ','A'),(4,'A ','A');
Query OK, 4 rows affected (0.00 sec)
Records: 4  Duplicates: 0  Warnings: 0

mysql> select count(distinct v1,v2) from t2;
+-----------------------+
| count(distinct v1,v2) |
+-----------------------+
|                     1 |
+-----------------------+
1 row in set (0.01 sec)

mysql>
mysql>
mysql>
mysql>
mysql> select count(distinct v1,v2) from t2;
+-----------------------+
| count(distinct v1,v2) |
+-----------------------+
|                     1 |
+-----------------------+
1 row in set (0.00 sec)

mysql> select * from t2;
+------+------+------+
| id   | v1   | v2   |
+------+------+------+
|    1 | a    | A    |
|    2 | A    | A    |
|    3 | a    | A    |
|    4 | A    | A    |
+------+------+------+
4 rows in set (0.00 sec)
```

### 2. What did you expect to see? (Required)
```
mysql> select count(distinct v1,v2) from t2;
+-----------------------+
| count(distinct v1,v2) |
+-----------------------+
|                     1 |
+-----------------------+
1 row in set (0.00 sec)
```
### 3. What did you see instead (Required)
```
mysql> select count(distinct v1,v2) from t2;
+-----------------------+
| count(distinct v1,v2) |
+-----------------------+
|                     2 |
+-----------------------+
1 row in set (0.00 sec)
```

### 4. What is your TiDB version? (Required)

master@a7fdc2a056636119202c16feefb91f25702a8f4f

### PR DESCRIPTION
**Pull request by windtalker**
### What problem does this PR solve?

Issue Number: close #27091  

Problem Summary:

As the issue described.
### What is changed and how it works?

Proposal: [xxx](url) 

What's Changed:

How it Works:

Use the collation info when calling `UpdatePartialResult` in `countOriginalWithDistinct` and `approxCountDistinctOriginal`
 
### Check List 

Tests 

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note 

```release-note
Fix bug that count disctinct on multi-columns return wrong result when new collation is on.
```

### CODE REVIEW

**Review by qw4990 — APPROVED** [review]

File: executor/aggfuncs/func_count_distinct.go [hunk 1]
```diff
@@ -401,6 +405,7 @@ func evalAndEncode(
 		if err != nil || isNull {
 			break
 		}
+		val = string(collator.Key(val))
 		encodedBytes = codec.EncodeCompactBytes(encodedBytes, hack.Slice(val))
```
Comment (wjhuang2016) [inline executor/aggfuncs/func_count_distinct.go:409]: codec.EncodeCompactBytes(encodedBytes, collator.Key(val))

Comment (wjhuang2016) [reply executor/aggfuncs/func_count_distinct.go:409]: No need to bytes -> string -> bytes

Comment (windtalker) [reply executor/aggfuncs/func_count_distinct.go]: OK, updated.

**Review by wjhuang2016 — APPROVED** [review]

**Pull request merged**